In [44]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer


# Modeling
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
from sklearn.metrics import accuracy_score, recall_score, precision_score, confusion_matrix

In [2]:
df = pd.read_pickle('../savedFiles/cleaned_df.pkl')

In [3]:
df.reset_index()

index                                              title  \
0        0  A Collapse That Wiped Out 5 Years of Growth, W...   
1        1  Grilled by Lawmakers, Big Tech Turns Up the Ga...   
2        2  How Pimco’s Cayman-Based Hedge Fund Can Profit...   
3        3          Big Tech Earnings Surge as Economy Slumps   
4        4  ‘This Is a New Phase’: Europe Shifts Tactics t...   
..     ...                                                ...   
799    799  BREAKING! This Revolutionary Cancer-Killing Tr...   
800    800  This Magic Mushroom Compound Will Help Treat D...   
801    801  This Magic Mushroom Compound Will Help Treat D...   
802    802  Trump Appoint Hollywood Legend To Be The Head ...   
803    803  Trump Appoint Hollywood Legend To Be The Head ...   

                                                  text  \
0    “Not only have we plateaued, but we may be los...   
1    “I don’t know the specifics of that situation,...   
2    WASHINGTON — Pacific Investment Management Com...   
3    Live Updated July 30, 2020, 8:22 p.m. ET July ...   
4    LONDON — European Union leaders are pursuing a...   
..                                                 ...   
799  This is the future of medicine. Novartis Pharm...   
800  Science is definitely amazing!\n\nNature is ab...   
801  Science is definitely amazing!\n\nNature is ab...   
802  Presiden Trump made serious changes in the Whi...   
803  Presiden Trump made serious changes in the Whi...   

                                                  link  \
0    https://www.nytimes.com/2020/07/30/business/ec...   
1    https://www.nytimes.com/2020/07/30/technology/...   
2    https://www.nytimes.com/2020/07/30/business/ec...   
3    https://www.nytimes.com/live/2020/07/30/busine...   
4    https://www.nytimes.com/2020/07/30/technology/...   
..                                                 ...   
799  https://www.usasupreme.com/breaking-revolution...   
800  https://www.usasupreme.com/magic-mushroom-comp...   
801  https://www.usasupreme.com/magic-mushroom-comp...   
802  https://www.usasupreme.com/trump-appoint-holly...   
803  https://www.usasupreme.com/trump-appoint-holly...   

                                                author  label  \
0                                      [Ben Casselman]      0   
1                                        [Kevin Roose]      0   
2                                     [Jeanna Smialek]      0   
3    [Mike Isaac, Karen Weise, Daisuke Wakabayashi,...      0   
4                                     [Adam Satariano]      0   
..                                                 ...    ...   
799  [Alex D., Alex D Is A Conservative Journalist,...      1   
800  [Natalie D., Natalie D. Is An American Conserv...      1   
801  [Natalie D., Natalie D. Is An American Conserv...      1   
802  [Alex D., Alex D Is A Conservative Journalist,...      1   
803  [Alex D., Alex D Is A Conservative Journalist,...      1   

                                           clean_title  \
0             collapse wiped years growth bounce sight   
1            grilled lawmakers big tech turns gaslight   
2      pimco cayman based hedge fund profit fed rescue   
3               big tech earnings surge economy slumps   
4     new phase europe shifts tactics limit tech power   
..                                                 ...   
799   breaking revolutionary cancer killing treatme...   
800      magic mushroom compound help treat depression   
801      magic mushroom compound help treat depression   
802   trump appoint hollywood legend head america f...   
803   trump appoint hollywood legend head america f...   

                                            clean_text  
0     plateaued may losing ground said diane swonk ...  
1     know specifics situation mr bezos replied ano...  
2     washington pacific investment management comp...  
3     live updated july p et july p et big tech ear...  
4     london european union leaders pursuing new la...  
..        

In [4]:
df['fake'] = df['label']

In [5]:
df

title  \
0    A Collapse That Wiped Out 5 Years of Growth, W...   
1    Grilled by Lawmakers, Big Tech Turns Up the Ga...   
2    How Pimco’s Cayman-Based Hedge Fund Can Profit...   
3            Big Tech Earnings Surge as Economy Slumps   
4    ‘This Is a New Phase’: Europe Shifts Tactics t...   
..                                                 ...   
799  BREAKING! This Revolutionary Cancer-Killing Tr...   
800  This Magic Mushroom Compound Will Help Treat D...   
801  This Magic Mushroom Compound Will Help Treat D...   
802  Trump Appoint Hollywood Legend To Be The Head ...   
803  Trump Appoint Hollywood Legend To Be The Head ...   

                                                  text  \
0    “Not only have we plateaued, but we may be los...   
1    “I don’t know the specifics of that situation,...   
2    WASHINGTON — Pacific Investment Management Com...   
3    Live Updated July 30, 2020, 8:22 p.m. ET July ...   
4    LONDON — European Union leaders are pursuing a...   
..                                                 ...   
799  This is the future of medicine. Novartis Pharm...   
800  Science is definitely amazing!\n\nNature is ab...   
801  Science is definitely amazing!\n\nNature is ab...   
802  Presiden Trump made serious changes in the Whi...   
803  Presiden Trump made serious changes in the Whi...   

                                                  link  \
0    https://www.nytimes.com/2020/07/30/business/ec...   
1    https://www.nytimes.com/2020/07/30/technology/...   
2    https://www.nytimes.com/2020/07/30/business/ec...   
3    https://www.nytimes.com/live/2020/07/30/busine...   
4    https://www.nytimes.com/2020/07/30/technology/...   
..                                                 ...   
799  https://www.usasupreme.com/breaking-revolution...   
800  https://www.usasupreme.com/magic-mushroom-comp...   
801  https://www.usasupreme.com/magic-mushroom-comp...   
802  https://www.usasupreme.com/trump-appoint-holly...   
803  https://www.usasupreme.com/trump-appoint-holly...   

                                                author  label  \
0                                      [Ben Casselman]      0   
1                                        [Kevin Roose]      0   
2                                     [Jeanna Smialek]      0   
3    [Mike Isaac, Karen Weise, Daisuke Wakabayashi,...      0   
4                                     [Adam Satariano]      0   
..                                                 ...    ...   
799  [Alex D., Alex D Is A Conservative Journalist,...      1   
800  [Natalie D., Natalie D. Is An American Conserv...      1   
801  [Natalie D., Natalie D. Is An American Conserv...      1   
802  [Alex D., Alex D Is A Conservative Journalist,...      1   
803  [Alex D., Alex D Is A Conservative Journalist,...      1   

                                           clean_title  \
0             collapse wiped years growth bounce sight   
1            grilled lawmakers big tech turns gaslight   
2      pimco cayman based hedge fund profit fed rescue   
3               big tech earnings surge economy slumps   
4     new phase europe shifts tactics limit tech power   
..                                                 ...   
799   breaking revolutionary cancer killing treatme...   
800      magic mushroom compound help treat depression   
801      magic mushroom compound help treat depression   
802   trump appoint hollywood legend head america f...   
803   trump appoint hollywood legend head america f...   

                                            clean_text  fake  
0     plateaued may losing ground said diane swonk ...     0  
1     know specifics situation mr bezos replied ano...     0  
2     washington pacific investment management comp...     0  
3     live updated july p et july p et big tech ear...     0  
4     london european union leaders pursuing new la...     0  
..                                                 ...   ...  
799   future medicine novartis pharmace

In [29]:
copy_df = df.copy()

In [30]:
copy_df

title  \
0    A Collapse That Wiped Out 5 Years of Growth, W...   
1    Grilled by Lawmakers, Big Tech Turns Up the Ga...   
2    How Pimco’s Cayman-Based Hedge Fund Can Profit...   
3            Big Tech Earnings Surge as Economy Slumps   
4    ‘This Is a New Phase’: Europe Shifts Tactics t...   
..                                                 ...   
799  BREAKING! This Revolutionary Cancer-Killing Tr...   
800  This Magic Mushroom Compound Will Help Treat D...   
801  This Magic Mushroom Compound Will Help Treat D...   
802  Trump Appoint Hollywood Legend To Be The Head ...   
803  Trump Appoint Hollywood Legend To Be The Head ...   

                                                  text  \
0    “Not only have we plateaued, but we may be los...   
1    “I don’t know the specifics of that situation,...   
2    WASHINGTON — Pacific Investment Management Com...   
3    Live Updated July 30, 2020, 8:22 p.m. ET July ...   
4    LONDON — European Union leaders are pursuing a...   
..                                                 ...   
799  This is the future of medicine. Novartis Pharm...   
800  Science is definitely amazing!\n\nNature is ab...   
801  Science is definitely amazing!\n\nNature is ab...   
802  Presiden Trump made serious changes in the Whi...   
803  Presiden Trump made serious changes in the Whi...   

                                                  link  \
0    https://www.nytimes.com/2020/07/30/business/ec...   
1    https://www.nytimes.com/2020/07/30/technology/...   
2    https://www.nytimes.com/2020/07/30/business/ec...   
3    https://www.nytimes.com/live/2020/07/30/busine...   
4    https://www.nytimes.com/2020/07/30/technology/...   
..                                                 ...   
799  https://www.usasupreme.com/breaking-revolution...   
800  https://www.usasupreme.com/magic-mushroom-comp...   
801  https://www.usasupreme.com/magic-mushroom-comp...   
802  https://www.usasupreme.com/trump-appoint-holly...   
803  https://www.usasupreme.com/trump-appoint-holly...   

                                                author  label  \
0                                      [Ben Casselman]      0   
1                                        [Kevin Roose]      0   
2                                     [Jeanna Smialek]      0   
3    [Mike Isaac, Karen Weise, Daisuke Wakabayashi,...      0   
4                                     [Adam Satariano]      0   
..                                                 ...    ...   
799  [Alex D., Alex D Is A Conservative Journalist,...      1   
800  [Natalie D., Natalie D. Is An American Conserv...      1   
801  [Natalie D., Natalie D. Is An American Conserv...      1   
802  [Alex D., Alex D Is A Conservative Journalist,...      1   
803  [Alex D., Alex D Is A Conservative Journalist,...      1   

                                           clean_title  \
0             collapse wiped years growth bounce sight   
1            grilled lawmakers big tech turns gaslight   
2      pimco cayman based hedge fund profit fed rescue   
3               big tech earnings surge economy slumps   
4     new phase europe shifts tactics limit tech power   
..                                                 ...   
799   breaking revolutionary cancer killing treatme...   
800      magic mushroom compound help treat depression   
801      magic mushroom compound help treat depression   
802   trump appoint hollywood legend head america f...   
803   trump appoint hollywood legend head america f...   

                                            clean_text  fake  
0     plateaued may losing ground said diane swonk ...     0  
1     know specifics situation mr bezos replied ano...     0  
2     washington pacific investment management comp...     0  
3     live updated july p et july p et big tech ear...     0  
4     london european union leaders pursuing new la...     0  
..                                                 ...   ...  
799   future medicine novartis pharmace

In [31]:
columns_to_delete = ['title','text','label','author','link', 'clean_title']

In [32]:
copy_df = copy_df.drop(columns_to_delete, axis = 1)

In [33]:
copy_df

clean_text  fake
0     plateaued may losing ground said diane swonk ...     0
1     know specifics situation mr bezos replied ano...     0
2     washington pacific investment management comp...     0
3     live updated july p et july p et big tech ear...     0
4     london european union leaders pursuing new la...     0
..                                                 ...   ...
799   future medicine novartis pharmaceuticals comp...     1
800   science definitely amazing nature abundant tr...     1
801   science definitely amazing nature abundant tr...     1
802   presiden trump made serious changes white hou...     1
803   presiden trump made serious changes white hou...     1

[804 rows x 2 columns]

In [34]:
copy_df['fake'].value_counts(normalize=True)

0    0.733831
1    0.266169
Name: fake, dtype: float64

In [35]:
X = df['clean_text']
y = df['fake']

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42)

In [39]:
X_train.shape

(643,)

In [42]:
X_test.shape

(161,)

## Model 1 - Countvectorizer and MultinomialNB

In [46]:
pipe = Pipeline([('cvec', CountVectorizer()),    
                 ('nb', MultinomialNB())])

# Tune GridSearchCV
pipe_params = {'cvec__ngram_range': [(1,1),(1,3)],
               'nb__alpha': [.36, .6]}

gs = GridSearchCV(pipe, param_grid=pipe_params, cv=3)
gs.fit(X_train, y_train);
print("Best score:", gs.best_score_)
print("Train score", gs.score(X_train, y_train))
print("Test score", gs.score(X_test, y_test))

gs.best_params_

Best score: 0.9128957473013113
Train score 0.9968895800933126
Test score 0.9440993788819876


{'cvec__ngram_range': (1, 1), 'nb__alpha': 0.6}

## Model 2 - tfidfvectorizer and MultinomialNB

In [47]:
pipe = Pipeline([('tvect', TfidfVectorizer()),    
                 ('nb', MultinomialNB())])

# Tune GridSearchCV
pipe_params = {'tvect__max_df': [.75, .98],
               'tvect__min_df': [4, 5],
               'tvect__ngram_range': [(1,2), (1,3)],
               'nb__alpha': [0.1, 1]}

gs = GridSearchCV(pipe, param_grid=pipe_params, cv=5)
gs.fit(X_train, y_train);
print("Best score:", gs.best_score_)
print("Train score", gs.score(X_train, y_train))
print("Test score", gs.score(X_test, y_test))

gs.best_params_

Best score: 0.9175024224806201
Train score 0.9906687402799378
Test score 0.9440993788819876


{'nb__alpha': 0.1,
 'tvect__max_df': 0.75,
 'tvect__min_df': 4,
 'tvect__ngram_range': (1, 2)}

## Model 3 - tfidfvectorizer and logistic regression

In [48]:
pipe = Pipeline([('tvect', TfidfVectorizer()),    
                 ('lr', LogisticRegression(solver='liblinear'))])

# Tune GridSearchCV
pipe_params = {'tvect__max_df': [.75, .98, 1.0],
               'tvect__min_df': [2, 3, 5],
               'tvect__ngram_range': [(1,1), (1,2), (1,3)],
               'lr__C': [1]}

gs = GridSearchCV(pipe, param_grid=pipe_params, cv=3)
gs.fit(X_train, y_train);
print("Best score:", gs.best_score_)
print("Train score", gs.score(X_train, y_train))
print("Test score", gs.score(X_test, y_test))

gs.best_params_

Best score: 0.8242700862131421
Train score 0.9129082426127527
Test score 0.8385093167701864


{'lr__C': 1,
 'tvect__max_df': 0.75,
 'tvect__min_df': 5,
 'tvect__ngram_range': (1, 3)}

## Model 4 - countvectorizer and LogisticRegression

In [50]:

pipe = Pipeline([('cvec', CountVectorizer()),    
                 ('lr', LogisticRegression(solver='liblinear'))])

# Tune GridSearchCV
pipe_params = {'cvec__ngram_range': [(1,1), (2,2), (1,3)],
               'lr__C': [0.01, 1]}

gs = GridSearchCV(pipe, param_grid=pipe_params, cv=3)
gs.fit(X_train, y_train);
print("Best score:", gs.best_score_)
print("Train score", gs.score(X_train, y_train))
print("Test score", gs.score(X_test, y_test))

gs.best_params_

Best score: 0.919126276896327
Train score 1.0
Test score 0.9254658385093167


{'cvec__ngram_range': (1, 1), 'lr__C': 1}

## as we can see multinomialNB model perform better than Logistic Regression

### though the tfidfvectorizer and countvectorizer have almost similer output 
best score is highest for tfidf vectorizer 

# Best Model

### tfidfvectorizer and MultinomialNB - Best Score

In [107]:
nb = MultinomialNB(alpha = 0.1)
tfidf = TfidfVectorizer( min_df = 5, max_df = 0.75)

tfidf.fit(X_train)
X_tfidf_train = tfidf.transform(X_train)
X_tfidf_test = tfidf.transform(X_test)

In [108]:
nb.fit(X_tfidf_train,y_train)

MultinomialNB(alpha=0.1)

In [109]:
preds = nb.predict(X_tfidf_test)

In [110]:
print(nb.score(X_tfidf_test, y_test))

0.9440993788819876


In [111]:
# Create a confusion matrix
cnf_matrix = metrics.confusion_matrix(y_test, preds)
cnf_matrix

array([[114,   3],
       [  6,  38]], dtype=int64)

In [112]:
print("Accuracy:",round(metrics.accuracy_score(y_test, preds)*100, 2),'%')
print("Precision:",round(metrics.precision_score(y_test, preds)*100, 2), '%')
print("Recall:",round(metrics.recall_score(y_test, preds)*100, 2), '%')
print("F1 score:",round(metrics.f1_score(y_test, preds)*100, 2), '%')

Accuracy: 94.41 %
Precision: 92.68 %
Recall: 86.36 %
F1 score: 89.41 %


# Most common Spam Words

In [137]:
import nltk
from collections import Counter
spam = df[df['fake']==1]

In [138]:
spam_words = nltk.word_tokenize(" ".join(spam['clean_text'].values.tolist()))

In [139]:
spam_counter = Counter(spam_words)
print(spam_counter.most_common(50))

[('trump', 447), ('people', 415), ('said', 366), ('president', 341), ('one', 313), ('us', 262), ('america', 253), ('new', 235), ('com', 224), ('https', 222), ('world', 217), ('would', 206), ('also', 198), ('obama', 198), ('two', 189), ('see', 185), ('like', 173), ('news', 173), ('time', 172), ('american', 168), ('police', 164), ('state', 163), ('years', 160), ('law', 157), ('video', 156), ('could', 156), ('many', 150), ('states', 149), ('earth', 133), ('first', 129), ('black', 129), ('going', 126), ('government', 126), ('country', 126), ('according', 126), ('fact', 124), ('www', 123), ('white', 123), ('know', 121), ('system', 121), ('get', 121), ('covid', 121), ('order', 118), ('great', 116), ('think', 115), ('health', 114), ('even', 114), ('media', 111), ('around', 110), ('god', 107)]


In [140]:
spam_bigrams = nltk.bigrams(spam_words)

In [141]:
spam_counter = Counter(spam_bigrams)
print(spam_counter.most_common(50))

[(('https', 'www'), 116), (('president', 'trump'), 88), (('united', 'states'), 75), (('donald', 'trump'), 70), (('abraham', 'isaac'), 55), (('isaac', 'jacob'), 55), (('god', 'abraham'), 53), (('black', 'lives'), 53), (('lives', 'matter'), 53), (('new', 'york'), 51), (('twitter', 'com'), 47), (('pic', 'twitter'), 44), (('american', 'people'), 42), (('fox', 'news'), 38), (('law', 'enforcement'), 38), (('world', 'order'), 37), (('readers', 'think'), 36), (('think', 'story'), 36), (('story', 'fact'), 36), (('fact', 'add'), 36), (('add', 'two'), 36), (('two', 'cents'), 36), (('holy', 'gematria'), 35), (('white', 'house'), 34), (('foreign', 'agents'), 32), (('https', 'co'), 32), (('around', 'world'), 31), (('new', 'world'), 31), (('baking', 'soda'), 30), (('years', 'ago'), 29), (('mainstream', 'media'), 28), (('get', 'best'), 28), (('humic', 'fulvic'), 27), (('christian', 'news'), 27), (('many', 'people'), 26), (('banned', 'foreign'), 26), (('reloaded', 'org'), 26), (('www', 'brighteon'), 25

 # Most common ham Words

In [148]:
ham = df[df['fake']==0]

In [149]:
ham_words = nltk.word_tokenize(" ".join(ham['clean_text'].values.tolist()))

In [150]:
ham_counter = Counter(ham_words)
print(ham_counter.most_common(50))

[('said', 1536), ('trump', 1299), ('people', 1256), ('one', 1185), ('new', 959), ('would', 869), ('time', 868), ('also', 865), ('r', 733), ('like', 725), ('election', 700), ('even', 689), ('state', 670), ('percent', 615), ('many', 610), ('year', 604), ('polls', 575), ('first', 560), ('think', 526), ('two', 519), ('president', 510), ('could', 510), ('get', 474), ('states', 473), ('us', 470), ('make', 468), ('business', 462), ('ad', 444), ('pandemic', 433), ('company', 428), ('may', 424), ('right', 421), ('way', 412), ('much', 407), ('going', 389), ('work', 388), ('need', 388), ('health', 386), ('covid', 385), ('data', 381), ('know', 377), ('country', 376), ('coronavirus', 371), ('still', 369), ('week', 361), ('team', 360), ('years', 358), ('back', 356), ('want', 351), ('day', 350)]


In [151]:

ham_bigrams = nltk.bigrams(ham_words)

In [152]:
ham_counter = Counter(ham_bigrams)
print(ham_counter.most_common(50))

[(('new', 'york'), 178), (('second', 'quarter'), 139), (('united', 'states'), 136), (('donald', 'trump'), 126), (('anne', 'applebaum'), 111), (('ad', 'ad'), 109), (('min', 'read'), 100), (('president', 'trump'), 96), (('july', 'min'), 94), (('health', 'care'), 93), (('michelle', 'goldberg'), 87), (('black', 'lives'), 85), (('lives', 'matter'), 85), (('mr', 'trump'), 84), (('york', 'times'), 83), (('white', 'house'), 81), (('hong', 'kong'), 80), (('even', 'though'), 75), (('r', 'r'), 74), (('frank', 'bruni'), 72), (('cooper', 'dyke'), 72), (('pollster', 'ratings'), 72), (('coronavirus', 'pandemic'), 71), (('year', 'old'), 70), (('percentage', 'points'), 70), (('plus', 'minus'), 70), (('read', 'opinions'), 69), (('opinions', 'expressed'), 69), (('expressed', 'entrepreneur'), 69), (('entrepreneur', 'contributors'), 69), (('social', 'media'), 68), (('last', 'week'), 68), (('trump', 'administration'), 68), (('joe', 'biden'), 68), (('public', 'health'), 68), (('child', 'care'), 64), (('avera

## Conclusions and Next-Steps

The most model to optimize for accuracy in detecting fake news and absurd news uses TfidfVectorizer and MultinomialDB. The optimal parameters for this model are where ngram_range = (1,2) and alpha = 0.1.

Accuracy: 94.41 %
Precision: 92.68 %
Recall: 86.36 %
F1 score: 89.41 %


